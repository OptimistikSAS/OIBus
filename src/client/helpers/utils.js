const dynamicSort = (property) => {
  let prop = property
  let sortOrder = 1
  if (prop[0] === '-') {
    sortOrder = -1
    prop = prop.substr(1)
  }
  return (a, b) => {
    let result
    const valueA = a[prop].toString()
    const valueB = b[prop].toString()
    if (valueA < valueB) {
      result = -1
    } else if (valueA > valueB) {
      result = 1
    } else {
      result = 0
    }
    return result * sortOrder
  }
}

export default { dynamicSort }
