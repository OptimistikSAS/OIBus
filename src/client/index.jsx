import React from 'react'
import ReactDOM from 'react-dom'
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom'
import { Container } from 'reactstrap'

import './style/main.less'

import TopHeader from './TopHeader.jsx'
import Activation from './Activation/Activation.jsx'
import NotFound from './NotFound.jsx'
import South from './South/South.jsx'
import North from './North/North.jsx'
import Engine from './Engine/Engine.jsx'
import ConfigureApi from './North/ConfigureApi.jsx'
import ConfigureProtocol from './South/ConfigureProtocol.jsx'
import ConfigurePoints from './South/ConfigurePoints.jsx'
import Logs from './Logs/Logs.jsx'
import Health from './Health/Health.jsx'
import AlertContainer from './components/AlertContainer.jsx'
import { AlertProvider } from './context/AlertContext.jsx'
import { ConfigProvider } from './context/configContext.jsx'

const Main = () => (
  <Router>
    <>
      <ConfigProvider>
        <AlertProvider>
          <TopHeader />
          <Container className="oi-container-with-top-nav" fluid>
            <AlertContainer />
            <Switch>
              <Route exact path="/" component={Health} />
              <Route exact path="/engine" component={Engine} />
              <Route exact path="/south" component={South} />
              <Route exact path="/south/:dataSourceId" component={ConfigureProtocol} />
              <Route exact path="/south/:dataSourceId/points" component={ConfigurePoints} />
              <Route exact path="/north" component={North} />
              <Route exact path="/north/:applicationId" component={ConfigureApi} />
              <Route exact path="/log" component={Logs} />
              <Route exact path="/activation" component={Activation} />
              <Route component={NotFound} />
            </Switch>
          </Container>
        </AlertProvider>
      </ConfigProvider>
    </>
  </Router>
)

ReactDOM.render(<Main />, document.getElementById('root'))
