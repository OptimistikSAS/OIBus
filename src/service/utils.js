const fs = require('node:fs/promises')
const { createReadStream, createWriteStream } = require('node:fs')
const zlib = require('node:zlib')
const path = require('node:path')

const minimist = require('minimist')
const { DateTime } = require('luxon')

const COMPRESSION_LEVEL = 9

/**
 * Get config file from console arguments
 * @returns {Object} - the config file and check argument
 */
const getCommandLineArguments = () => {
  const args = minimist(process.argv.slice(2))
  const { config = './oibus.json', check = false } = args
  return { configFile: path.resolve(config), check }
}

/**
 * Method to return a delayed promise.
 * @param {Number} timeout - The delay
 * @return {Promise<any>} - The delay promise
 */
const delay = async (timeout) => new Promise((resolve) => {
  setTimeout(resolve, timeout)
})

/**
 * Compute a list of end interval from a startTime, endTime and maxInterval.
 * @param {Date} startTime - The start of the interval
 * @param {Date} endTime - The end of the interval
 * @param {Number} maxInterval - The interval in s to split the [startTime - endTime] interval into
 * @returns {[{Date, Date}]} - The list of intervals
 */
const generateIntervals = (startTime, endTime, maxInterval) => {
  const originalInterval = endTime.getTime() - startTime.getTime()
  if (maxInterval > 0 && (endTime.getTime() - startTime.getTime()) > 1000 * maxInterval) {
    const numberOfInterval = originalInterval / (maxInterval * 1000)
    const intervalLists = []
    for (let i = 0; i < numberOfInterval; i += 1) {
      // Compute the newStartTime and the newEndTime for each interval
      const newStartTime = new Date(startTime.getTime() + i * 1000 * maxInterval)
      const newEndTime = new Date(startTime.getTime() + (i + 1) * 1000 * maxInterval)

      // If the newEndTime is bigger than the original endTime, the definitive end of the interval must be endTime
      intervalLists.push({
        startTime: newStartTime,
        endTime: newEndTime < endTime ? newEndTime : endTime,
      })
    }
    return intervalLists
  }
  return [{ startTime, endTime }]
}

/**
 * Create a folder if it does not exist
 * @param {String} folder - The folder to create
 * @return {Promise<void>} - The result promise
 */
const createFolder = async (folder) => {
  const folderPath = path.resolve(folder)
  try {
    await fs.stat(folderPath)
  } catch (error) {
    await fs.mkdir(folderPath, { recursive: true })
  }
}

/**
 * Replace the variables such as @CurrentDate in the file name with their values
 * @param {String} filename - The filename to change
 * @param {Number} queryPart - The part of the query being executed
 * @param {String} connectorName - The name of the connector
 * @return {String} - The renamed file name
 */
const replaceFilenameWithVariable = (filename, queryPart, connectorName) => filename.replace('@CurrentDate', DateTime.local()
  .toFormat('yyyy_MM_dd_HH_mm_ss_SSS'))
  .replace('@ConnectorName', connectorName)
  .replace('@QueryPart', `${queryPart}`)

/**
 * Generate date based on the configured format taking into account the timezone configuration.
 * Ex: With timezone "Europe/Paris" the date "2019-01-01 00:00:00" will be converted to "Tue Jan 01 2019 00:00:00 GMT+0100"
 * @param {String} date - The date to parse and format
 * @param {String} timezone - The timezone to use to replace the timezone of the date
 * @param {String} dateFormat - The format of the date
 * @returns {String} - The formatted date with timezone
 */
const generateDateWithTimezone = (date, timezone, dateFormat) => {
  if (DateTime.fromISO(date).isValid) {
    return date
  }
  return DateTime.fromFormat(date, dateFormat, { zone: timezone }).toJSDate().toISOString()
}

/**
 * Compress the specified file
 * @param {String} input - The path of the file to compress
 * @param {String} output - The path to the compressed file
 * @returns {Promise<void>} - The result promise
 */
const compress = async (input, output) => new Promise((resolve, reject) => {
  const readStream = createReadStream(input)
  const writeStream = createWriteStream(output)
  const gzip = zlib.createGzip({ level: COMPRESSION_LEVEL })
  readStream
    .pipe(gzip)
    .pipe(writeStream)
    .on('error', (error) => {
      reject(error)
    })
    .on('finish', () => {
      resolve()
    })
})

/**
 * Check if a file exists in async way
 * @param {String} filePath - The file path to test
 * @returns {Promise<boolean>} - If the file exists or not
 */
const filesExists = async (filePath) => {
  try {
    await fs.stat(filePath)
  } catch {
    return false
  }
  return true
}

module.exports = {
  getCommandLineArguments,
  delay,
  generateIntervals,
  createFolder,
  replaceFilenameWithVariable,
  generateDateWithTimezone,
  compress,
  filesExists,
}
