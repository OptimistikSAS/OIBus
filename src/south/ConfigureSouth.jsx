import React from 'react'
import Form from 'react-jsonschema-form-bs4'
import { withRouter } from 'react-router-dom'
import { getScheme } from './Schemas'

// eslint-disable-next-line react/prop-types
const ConfigureSouth = ({ match, location }) => {
  const [configJson, setConfigJson] = React.useState()
  const [configSchema, setConfigSchema] = React.useState()

  const updateForm = () => {
    const { protocol } = match.params
    const { formData } = location
    setConfigJson(formData)
    setConfigSchema(getScheme(protocol))
  }
  React.useEffect(() => {
    updateForm()
  }, [])

  const handleChange = (data) => {
    const { formData } = data
    const { protocol } = formData

    if (configJson !== data.formData) {
      setConfigJson(formData)
      setConfigSchema(getScheme(protocol))
    }
  }
  const log = type => console.info.bind(console, type)
  return (
    <>
      {configJson && configSchema && (
        <Form
          formData={configJson}
          liveValidate
          schema={configSchema}
          // uiSchema={ConfigureSouth.uiModbus}
          autocomplete="on"
          onChange={handleChange}
          onSubmit={log('submitted')}
          onError={log('errors')}
        />
      )}
      <pre>{configJson && JSON.stringify(configJson, ' ', 2)}</pre>
    </>
  )
}

export default withRouter(ConfigureSouth)

ConfigureSouth.schema = {
  title: 'South',
  type: 'object',
  properties: {
    Modbus: {
      type: 'object',
      title: 'Modbus',
      properties: {
        addressGap: {
          type: 'object',
          title: 'Address Gap',
          properties: {
            holdingRegister: { type: 'number', title: 'Holding register' },
            coil: { type: 'number', title: 'Coil' },
          },
        },
      },
    },
    equipments: {
      type: 'array',
      title: 'Equipments',
      items: {
        type: 'object',
        properties: {
          equipmentId: {
            type: 'string',
            title: 'Equipment ID',
          },
          enabled: {
            type: 'boolean',
            title: 'Enabled',
            default: true,
          },
          protocol: {
            type: 'string',
            enum: ['CSV', 'MQTT', 'OPCUA', 'RawFile', 'Modbus'],
            title: 'Protocol',
            default: 'CSV',
          },
          // pointIdRoot: {
          //   type: 'string',
          //   title: 'Point ID Root',
          // },
          // defaultScanMode: {
          //   type: 'string',
          //   title: 'Default Scan Mode',
          //   default: 'every20Second',
          // },
          // '^[A-Z]+$': {
          //   type: 'object',
          //   properties: {
          //     inputFolder: {
          //       type: 'string',
          //       title: 'Input Folder',
          //     },
          //     archiveFolder: {
          //       type: 'string',
          //       title: 'Archive Folder',
          //     },
          //     errorFolder: {
          //       type: 'string',
          //       title: 'Error Folder',
          //     },
          //     separator: {
          //       type: 'string',
          //       title: 'Separator',
          //       default: ',',
          //     },
          //     timeColumn: {
          //       type: 'number',
          //       title: 'Time Column',
          //       default: 0,
          //     },
          //     hasFirstLine: {
          //       type: 'boolean',
          //       title: 'Has First Line',
          //       default: true,
          //     },
          //   },
          // },
          // points: {
          //   type: 'array',
          //   title: 'Points',
          //   items: {},
          // },
        },
      },
    },
  },
}
