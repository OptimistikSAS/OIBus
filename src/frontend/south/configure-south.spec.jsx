/**
 * @jest-environment jsdom
 */
import React from 'react'
import { act, Simulate } from 'react-dom/test-utils'
import * as ReactDOMClient from 'react-dom/client'
import { testConfig } from '../../../tests/test-config.js'
import ConfigureSouth from './configure-south.jsx'
import utils from '../helpers/utils.js'

const dispatchNewConfig = jest.fn()
React.useContext = jest.fn().mockReturnValue({ newConfig: testConfig, dispatchNewConfig })
jest.mock('react-router-dom', () => (
  {
    useParams: jest.fn().mockReturnValue({ id: 'south-uuid-1' }),
    useNavigate: jest.fn(),
  }
))

const mockMath = Object.create(global.Math)
mockMath.random = () => 1
global.Math = mockMath

let container
let root
// eslint-disable-next-line no-undef
globalThis.IS_REACT_ACT_ENVIRONMENT = true
beforeEach(() => {
  container = document.createElement('div')
  root = ReactDOMClient.createRoot(container)
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
  root = null
})

describe('ConfigureSouth', () => {
  test('check ConfigureSouth', () => {
    act(() => {
      root.render(<ConfigureSouth />)
    })
    expect(container).toMatchSnapshot()
  })
  test('check update', () => {
    act(() => {
      root.render(<ConfigureSouth />)
    })
    act(() => {
      Simulate.change(document.getElementById('south.0.settings.inputFolder'), { target: { value: './myNewInputFolder' } })
    })

    expect(dispatchNewConfig).toBeCalledWith({
      type: 'update',
      name: 'south.0.settings.inputFolder',
      value: './myNewInputFolder',
      validity: null,
    })
    expect(container).toMatchSnapshot()
  })

  test('check south not found', () => {
    const reactUseContextMock = React.useContext
    // temporary empty array
    const config = utils.jsonCopy(testConfig)
    config.south = []
    React.useContext = jest.fn().mockReturnValue({ newConfig: config, dispatchNewConfig })
    act(() => {
      root.render(<ConfigureSouth />)
    })
    expect(container).toMatchSnapshot()
    React.useContext = reactUseContextMock
  })
})
