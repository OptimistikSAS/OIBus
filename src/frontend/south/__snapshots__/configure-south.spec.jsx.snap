// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigureSouth check ConfigureSouth 1`] = `
<div>
  <div
    class="d-flex align-items-center w-100 oi-sub-nav"
  >
    <h6
      class="text-muted d-flex align-items-center ps-3 pt-2 pb-2 mb-0"
    >
      <div>
        TestFolderScanner
      </div>
      <svg
        class="oi-icon mx-2"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
        />
      </svg>
    </h6>
    <div
      class="pull-right me-3"
    >
      <button
        class="inline-button autosize oi-status-button btn btn-outline-secondary btn-sm"
        type="button"
      >
        Status
      </button>
      <button
        class="inline-button autosize oi-points-button btn btn-secondary btn-sm disabled"
        disabled=""
      >
        File
      </button>
    </div>
  </div>
  <div
    class="container-fluid"
  >
    <form
      class="m-2"
    >
      <div
        class="row"
      >
        <h5>
          General settings
          <button
            class="util-button mt-1 btn btn-link"
            id="id"
          >
            <svg
              class="oi-help"
              fill="currentColor"
              height="12"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
              />
            </svg>
          </button>
        </h5>
      </div>
      <div
        class="container-fluid"
      >
        <div
          class="collapse"
          toggler="id"
        >
          <div
            class="row"
            style="margin-bottom: 15px;"
          >
            <div
              class="col"
            >
              <ul>
                <li>
                  This form allows to configure South-specific parameters.
                </li>
                <li>
                  You need to activate the South connector with the enabled checkbox.
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Disabled
            </label>
            <input
              class="oi-form-input checkbox form-check-input"
              id="south.0.enabled"
              name="south.0.enabled"
              type="checkbox"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                Enable this South connector
              </div>
            </small>
          </div>
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="mb-3"
          >
            <select
              aria-invalid="false"
              class="oi-form-input form-select"
              id="south.0.scanMode"
              name="south.0.scanMode"
            >
              <option
                value="everySecond"
              >
                everySecond
              </option>
              <option
                value="every10Second"
              >
                every10Second
              </option>
              <option
                value="every1Min"
              >
                every1Min
              </option>
              <option
                value="listen"
              >
                listen
              </option>
            </select>
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-12"
        >
          <div
            class="row"
          >
            <h5>
              FolderScanner settings
              <button
                class="util-button mt-1 btn btn-link"
                id="id"
              >
                <svg
                  class="oi-help"
                  fill="currentColor"
                  height="12"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="12"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                  />
                </svg>
              </button>
            </h5>
          </div>
          <div
            class="container-fluid"
          >
            <div
              class="collapse"
              toggler="id"
            >
              <div
                class="row"
                style="margin-bottom: 15px;"
              >
                <div
                  class="col"
                >
                  <p>
                    The FolderScanner South periodically checks the input folder for new files at an interval specified by scan mode. When a new file is detected it will be sent to any North capable of handling files and configured to accept files from this South.
                  </p>
                  <p>
                    Please use always the / separator. OIBus will convert it according to the platform. For example, on Windows, c:/input/ will be converted to c:\\input\\ Make sure OIBus has READ and WRITE access to this folder.
                    <b>
                      When OIBus is started as a Windows service, you may have to assign a user logon in the Windows services.msc console.
                    </b>
                  </p>
                  <p>
                    It is possible to specify how to handle files.
                  </p>
                  <ul>
                    <li>
                      If Preserve File is checked the files will be left intact, otherwise they will be removed from the input folder.
                    </li>
                    <li>
                      If Ignore Modified Date is checked the last edited time of the file will be ignored when the file is preserved. This implies that a file will be sent at each scan of the folder
                    </li>
                    <li>
                      The Minimum Age option is used to specify the time since the file was last modified. It is used to prevent handling files that are still populated.
                    </li>
                    <li>
                      The RegExp option can be used to filter what kind of files to be handled.
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="south.0.settings.inputFolder"
            >
              Input folder
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="south.0.settings.inputFolder"
              name="south.0.settings.inputFolder"
              type="text"
              value="./input/"
            />
            <div
              class="invalid-feedback"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                Path to folder such as: c:/input/
              </div>
            </small>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Preserve File?
            </label>
            <input
              checked=""
              class="oi-form-input checkbox form-check-input"
              id="south.0.settings.preserveFiles"
              name="south.0.settings.preserveFiles"
              type="checkbox"
            />
          </div>
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Ignore modified date
            </label>
            <input
              class="oi-form-input checkbox form-check-input"
              id="south.0.settings.ignoreModifiedDate"
              name="south.0.settings.ignoreModifiedDate"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="south.0.settings.minAge"
            >
              Minimum Age
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="south.0.settings.minAge"
              name="south.0.settings.minAge"
              type="integer"
              value="0"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="south.0.settings.regex"
            >
              RegExp
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="south.0.settings.regex"
              name="south.0.settings.regex"
              type="text"
              value=".csv"
            />
            <div
              class="invalid-feedback"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                RegExp to filter the folder
              </div>
            </small>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Compress File?
            </label>
            <input
              class="oi-form-input checkbox form-check-input"
              id="south.0.settings.compression"
              name="south.0.settings.compression"
              type="checkbox"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`ConfigureSouth check south not found 1`] = `
<div>
  <div
    class="spinner-container"
  >
    <div
      class="spinner-grow text-primary"
      role="status"
    >
      <span
        class="visually-hidden"
      >
        Loading...
      </span>
    </div>
    ...loading configuration from OIBus server...
  </div>
</div>
`;

exports[`ConfigureSouth check update 1`] = `
<div>
  <div
    class="d-flex align-items-center w-100 oi-sub-nav"
  >
    <h6
      class="text-muted d-flex align-items-center ps-3 pt-2 pb-2 mb-0"
    >
      <div>
        TestFolderScanner
      </div>
      <svg
        class="oi-icon mx-2"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
        />
      </svg>
    </h6>
    <div
      class="pull-right me-3"
    >
      <button
        class="inline-button autosize oi-status-button btn btn-outline-secondary btn-sm"
        type="button"
      >
        Status
      </button>
      <button
        class="inline-button autosize oi-points-button btn btn-secondary btn-sm disabled"
        disabled=""
      >
        File
      </button>
    </div>
  </div>
  <div
    class="container-fluid"
  >
    <form
      class="m-2"
    >
      <div
        class="row"
      >
        <h5>
          General settings
          <button
            class="util-button mt-1 btn btn-link"
            id="id"
          >
            <svg
              class="oi-help"
              fill="currentColor"
              height="12"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
              />
            </svg>
          </button>
        </h5>
      </div>
      <div
        class="container-fluid"
      >
        <div
          class="collapse"
          toggler="id"
        >
          <div
            class="row"
            style="margin-bottom: 15px;"
          >
            <div
              class="col"
            >
              <ul>
                <li>
                  This form allows to configure South-specific parameters.
                </li>
                <li>
                  You need to activate the South connector with the enabled checkbox.
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Disabled
            </label>
            <input
              class="oi-form-input checkbox form-check-input"
              id="south.0.enabled"
              name="south.0.enabled"
              type="checkbox"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                Enable this South connector
              </div>
            </small>
          </div>
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="mb-3"
          >
            <select
              aria-invalid="false"
              class="oi-form-input form-select"
              id="south.0.scanMode"
              name="south.0.scanMode"
            >
              <option
                value="everySecond"
              >
                everySecond
              </option>
              <option
                value="every10Second"
              >
                every10Second
              </option>
              <option
                value="every1Min"
              >
                every1Min
              </option>
              <option
                value="listen"
              >
                listen
              </option>
            </select>
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-12"
        >
          <div
            class="row"
          >
            <h5>
              FolderScanner settings
              <button
                class="util-button mt-1 btn btn-link"
                id="id"
              >
                <svg
                  class="oi-help"
                  fill="currentColor"
                  height="12"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="12"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                  />
                </svg>
              </button>
            </h5>
          </div>
          <div
            class="container-fluid"
          >
            <div
              class="collapse"
              toggler="id"
            >
              <div
                class="row"
                style="margin-bottom: 15px;"
              >
                <div
                  class="col"
                >
                  <p>
                    The FolderScanner South periodically checks the input folder for new files at an interval specified by scan mode. When a new file is detected it will be sent to any North capable of handling files and configured to accept files from this South.
                  </p>
                  <p>
                    Please use always the / separator. OIBus will convert it according to the platform. For example, on Windows, c:/input/ will be converted to c:\\input\\ Make sure OIBus has READ and WRITE access to this folder.
                    <b>
                      When OIBus is started as a Windows service, you may have to assign a user logon in the Windows services.msc console.
                    </b>
                  </p>
                  <p>
                    It is possible to specify how to handle files.
                  </p>
                  <ul>
                    <li>
                      If Preserve File is checked the files will be left intact, otherwise they will be removed from the input folder.
                    </li>
                    <li>
                      If Ignore Modified Date is checked the last edited time of the file will be ignored when the file is preserved. This implies that a file will be sent at each scan of the folder
                    </li>
                    <li>
                      The Minimum Age option is used to specify the time since the file was last modified. It is used to prevent handling files that are still populated.
                    </li>
                    <li>
                      The RegExp option can be used to filter what kind of files to be handled.
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="south.0.settings.inputFolder"
            >
              Input folder
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="south.0.settings.inputFolder"
              name="south.0.settings.inputFolder"
              type="text"
              value="./input/"
            />
            <div
              class="invalid-feedback"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                Path to folder such as: c:/input/
              </div>
            </small>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Preserve File?
            </label>
            <input
              checked=""
              class="oi-form-input checkbox form-check-input"
              id="south.0.settings.preserveFiles"
              name="south.0.settings.preserveFiles"
              type="checkbox"
            />
          </div>
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Ignore modified date
            </label>
            <input
              class="oi-form-input checkbox form-check-input"
              id="south.0.settings.ignoreModifiedDate"
              name="south.0.settings.ignoreModifiedDate"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="south.0.settings.minAge"
            >
              Minimum Age
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="south.0.settings.minAge"
              name="south.0.settings.minAge"
              type="integer"
              value="0"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="south.0.settings.regex"
            >
              RegExp
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="south.0.settings.regex"
              name="south.0.settings.regex"
              type="text"
              value=".csv"
            />
            <div
              class="invalid-feedback"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                RegExp to filter the folder
              </div>
            </small>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div
            class="mb-3 form-check form-switch"
          >
            <label
              class="form-check-label"
            >
              Compress File?
            </label>
            <input
              class="oi-form-input checkbox form-check-input"
              id="south.0.settings.compression"
              name="south.0.settings.compression"
              type="checkbox"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
